File: sdk/go/schema_service.go
package cratebase

// SchemaService manages collection schemas. Requires a superuser token in
// Client.AuthStore.
type SchemaService struct {
	client *Client
}

// GetList lists every collection.
func (s *SchemaService) GetList() ([]CollectionModel, error) {
	var result []CollectionModel
	err := s.client.Send("/api/collections", SendOptions{Method: "GET"}, &result)
	return result, err
}

// GetOne fetches a single collection by id or name.
func (s *SchemaService) GetOne(idOrName string) (*CollectionModel, error) {
	var result CollectionModel
	err := s.client.Send("/api/collections/"+idOrName, SendOptions{Method: "GET"}, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Create creates a new collection.
func (s *SchemaService) Create(data CollectionModel) (*CollectionModel, error) {
	var result CollectionModel
	err := s.client.Send("/api/collections", SendOptions{Method: "POST", Body: data}, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Update updates a collection's schema/rules by id or name.
func (s *SchemaService) Update(idOrName string, data CollectionModel) (*CollectionModel, error) {
	var result CollectionModel
	err := s.client.Send("/api/collections/"+idOrName, SendOptions{Method: "PATCH", Body: data}, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Delete deletes a collection by id or name.
func (s *SchemaService) Delete(idOrName string) error {
	return s.client.Send("/api/collections/"+idOrName, SendOptions{Method: "DELETE"}, nil)
}

File: sdk/python/src/cratebase/schema_service.py
"""Manage collection schemas. Requires a superuser token in `client.auth_store`."""

from __future__ import annotations

from typing import Any, Dict, List, TYPE_CHECKING

from .types import CollectionModel

if TYPE_CHECKING:
    from .client import Cratebase


class SchemaService:
    def __init__(self, client: "Cratebase") -> None:
        self._client = client

    def get_list(self) -> List[CollectionModel]:
        return self._client.send("/api/collections", method="GET")

    def get_one(self, id_or_name: str) -> CollectionModel:
        return self._client.send(f"/api/collections/{id_or_name}", method="GET")

    def create(self, data: Dict[str, Any]) -> CollectionModel:
        return self._client.send("/api/collections", method="POST", body=data)

    def update(self, id_or_name: str, data: Dict[str, Any]) -> CollectionModel:
        return self._client.send(f"/api/collections/{id_or_name}", method="PATCH", body=data)

    def delete(self, id_or_name: str) -> None:
        self._client.send(f"/api/collections/{id_or_name}", method="DELETE")
